File: backend/internal/domain/services/transaction_service.go
package services

import (
	"errors"
	"strings"
	"time"

	"github.com/quocdaijr/finance-management-backend/internal/domain/models"
	"github.com/quocdaijr/finance-management-backend/internal/repository"
	"gorm.io/gorm"
)

// TransactionService handles business logic for transactions
type TransactionService struct {
	transactionRepo *repository.TransactionRepository
	accountRepo     *repository.AccountRepository
	db              *gorm.DB
}

// NewTransactionService creates a new transaction service
func NewTransactionService(
	transactionRepo *repository.TransactionRepository,
	accountRepo *repository.AccountRepository,
	db *gorm.DB,
) *TransactionService {
	return &TransactionService{
		transactionRepo: transactionRepo,
		accountRepo:     accountRepo,
		db:              db,
	}
}

// Create creates a new transaction atomically
func (s *TransactionService) Create(userID uint, req *models.TransactionRequest) (*models.Transaction, error) {
	var transaction *models.Transaction

	// Start database transaction
	err := s.db.Transaction(func(tx *gorm.DB) error {
		// Check if account exists and belongs to user
		account, err := s.accountRepo.GetByID(req.AccountID, userID)
		if err != nil {
			return err
		}

		// Create transaction
		transaction = &models.Transaction{
			UserID:      userID,
			Amount:      req.Amount,
			Description: req.Description,
			Category:    req.Category,
			Type:        req.Type,
			Date:        req.Date,
			AccountID:   account.ID,
			Tags:        strings.Join(req.Tags, ","),
		}

		// Save transaction
		if err := tx.Create(transaction).Error; err != nil {
			return err
		}

		// Update account balance
		if req.Type == "income" {
			account.Balance += req.Amount
		} else {
			account.Balance -= req.Amount
		}

		// Save account with updated balance
		if err := tx.Save(account).Error; err != nil {
			return err
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	return transaction, nil
}

// GetByID gets a transaction by ID
func (s *TransactionService) GetByID(id uint, userID uint) (*models.Transaction, error) {
	return s.transactionRepo.GetByID(id, userID)
}

// GetAll gets all transactions for a user
func (s *TransactionService) GetAll(userID uint) ([]models.Transaction, error) {
	return s.transactionRepo.GetAll(userID)
}

// GetFiltered gets transactions with filters and pagination
func (s *TransactionService) GetFiltered(userID uint, filter *models.TransactionFilter) (*models.PaginatedTransactionResponse, error) {
	transactions, total, err := s.transactionRepo.GetFiltered(userID, filter)
	if err != nil {
		return nil, err
	}

	// Convert to response format
	var data []*models.TransactionResponse
	for _, t := range transactions {
		data = append(data, t.ToResponse())
	}

	// Calculate total pages
	pageSize := filter.PageSize
	if pageSize < 1 {
		pageSize = 20
	}
	totalPages := int(total) / pageSize
	if int(total)%pageSize > 0 {
		totalPages++
	}

	page := filter.Page
	if page < 1 {
		page = 1
	}

	return &models.PaginatedTransactionResponse{
		Data:       data,
		Total:      total,
		Page:       page,
		PageSize:   pageSize,
		TotalPages: totalPages,
	}, nil
}

// Update updates a transaction atomically
func (s *TransactionService) Update(id uint, userID uint, req *models.TransactionRequest) (*models.Transaction, error) {
	var updatedTransaction *models.Transaction

	// Start database transaction
	err := s.db.Transaction(func(tx *gorm.DB) error {
		// Get existing transaction
		transaction, err := s.transactionRepo.GetByID(id, userID)
		if err != nil {
			return err
		}

		// Check if account exists and belongs to user
		newAccount, err := s.accountRepo.GetByID(req.AccountID, userID)
		if err != nil {
			return err
		}

		// Get old account if different
		var oldAccount *models.Account
		if transaction.AccountID != newAccount.ID {
			oldAccount, err = s.accountRepo.GetByID(transaction.AccountID, userID)
			if err != nil {
				return err
			}
		}

		// Calculate balance adjustments
		if oldAccount != nil {
			// Reverse the effect of the old transaction on old account
			if transaction.Type == "income" {
				oldAccount.Balance -= transaction.Amount
			} else {
				oldAccount.Balance += transaction.Amount
			}

			// Save old account
			if err := tx.Save(oldAccount).Error; err != nil {
				return err
			}
		} else {
			// Same account, reverse old transaction first
			if transaction.Type == "income" {
				newAccount.Balance -= transaction.Amount
			} else {
				newAccount.Balance += transaction.Amount
			}
		}

		// Apply the new transaction
		if req.Type == "income" {
			newAccount.Balance += req.Amount
		} else {
			newAccount.Balance -= req.Amount
		}

		// Save new account
		if err := tx.Save(newAccount).Error; err != nil {
			return err
		}

		// Update transaction
		transaction.Amount = req.Amount
		transaction.Description = req.Description
		transaction.Category = req.Category
		transaction.Type = req.Type
		transaction.Date = req.Date
		transaction.AccountID = newAccount.ID
		transaction.Tags = strings.Join(req.Tags, ",")

		// Save transaction
		if err := tx.Save(transaction).Error; err != nil {
			return err
		}

		updatedTransaction = transaction
		return nil
	})

	if err != nil {
		return nil, err
	}

	return updatedTransaction, nil
}

// Delete deletes a transaction atomically
func (s *TransactionService) Delete(id uint, userID uint) error {
	// Start database transaction
	return s.db.Transaction(func(tx *gorm.DB) error {
		// Get transaction
		transaction, err := s.transactionRepo.GetByID(id, userID)
		if err != nil {
			return err
		}

		// Get account
		account, err := s.accountRepo.GetByID(transaction.AccountID, userID)
		if err != nil {
			return err
		}

		// Adjust account balance (reverse the transaction)
		if transaction.Type == "income" {
			account.Balance -= transaction.Amount
		} else {
			account.Balance += transaction.Amount
		}

		// Save account
		if err := tx.Save(account).Error; err != nil {
			return err
		}

		// Delete transaction
		if err := tx.Delete(&models.Transaction{}, id).Error; err != nil {
			return err
		}

		return nil
	})
}

// GetCategories gets all transaction categories
func (s *TransactionService) GetCategories() []string {
	return []string{
		"Food & Dining",
		"Shopping",
		"Housing",
		"Transportation",
		"Entertainment",
		"Health & Fitness",
		"Travel",
		"Education",
		"Personal Care",
		"Gifts & Donations",
		"Bills & Utilities",
		"Income",
		"Investments",
		"Uncategorized",
	}
}

// GetSummary gets a summary of transactions for a specific period
func (s *TransactionService) GetSummary(userID uint, period string) (*models.TransactionSummary, error) {
	// Calculate start and end dates based on period
	now := time.Now()
	var startDate time.Time

	switch period {
	case "week":
		startDate = now.AddDate(0, 0, -7)
	case "month":
		startDate = now.AddDate(0, -1, 0)
	case "year":
		startDate = now.AddDate(-1, 0, 0)
	default:
		startDate = now.AddDate(0, -1, 0)
	}

	return s.transactionRepo.GetSummary(userID, startDate, now)
}

// Transfer handles money transfer between two accounts atomically
func (s *TransactionService) Transfer(userID uint, req *models.TransferRequest) (*models.TransferResponse, error) {
	var response *models.TransferResponse

	// Validate that accounts are different
	if req.FromAccountID == req.ToAccountID {
		return nil, errors.New("cannot transfer to the same account")
	}

	// Start database transaction for atomic transfer
	err := s.db.Transaction(func(tx *gorm.DB) error {
		// Get source account
		fromAccount, err := s.accountRepo.GetByID(req.FromAccountID, userID)
		if err != nil {
			return errors.New("source account not found")
		}

		// Get destination account
		toAccount, err := s.accountRepo.GetByID(req.ToAccountID, userID)
		if err != nil {
			return errors.New("destination account not found")
		}

		// Check if source account has sufficient balance
		if fromAccount.Balance < req.Amount {
			return errors.New("insufficient balance in source account")
		}

		// Create description if not provided
		description := req.Description
		if description == "" {
			description = "Transfer from " + fromAccount.Name + " to " + toAccount.Name
		}

		// Create outgoing transaction (from source account)
		fromTransaction := &models.Transaction{
			UserID:      userID,
			Amount:      req.Amount,
			Description: description,
			Category:    "Transfer",
			Type:        "transfer",
			Date:        req.Date,
			AccountID:   fromAccount.ID,
			Tags:        strings.Join(req.Tags, ","),
		}

		if err := tx.Create(fromTransaction).Error; err != nil {
			return errors.New("failed to create outgoing transaction")
		}

		// Create incoming transaction (to destination account)
		toTransaction := &models.Transaction{
			UserID:      userID,
			Amount:      req.Amount,
			Description: description,
			Category:    "Transfer",
			Type:        "transfer",
			Date:        req.Date,
			AccountID:   toAccount.ID,
			Tags:        strings.Join(req.Tags, ","),
		}

		if err := tx.Create(toTransaction).Error; err != nil {
			return errors.New("failed to create incoming transaction")
		}

		// Update account balances
		fromAccount.Balance -= req.Amount
		toAccount.Balance += req.Amount

		if err := tx.Save(fromAccount).Error; err != nil {
			return errors.New("failed to update source account balance")
		}

		if err := tx.Save(toAccount).Error; err != nil {
			return errors.New("failed to update destination account balance")
		}

		response = &models.TransferResponse{
			FromTransaction: fromTransaction.ToResponse(),
			ToTransaction:   toTransaction.ToResponse(),
			Message:         "Transfer completed successfully",
		}

		return nil
	})

	if err != nil {
		return nil, err
	}

	return response, nil
}

File: frontend/src/services/transactionService.js
import Transaction from '../models/Transaction';
import httpClient from './httpClient';

/**
 * Transaction Service
 * Handles transaction-related operations through the Golang Gin backend API
 */
const transactionService = {
  /**
   * Get all transactions with pagination and filtering
   * @param {Object} filters - Optional filters
   * @param {number} filters.page - Page number (default: 1)
   * @param {number} filters.page_size - Page size (default: 20, max: 100)
   * @param {string} filters.start_date - Start date (YYYY-MM-DD)
   * @param {string} filters.end_date - End date (YYYY-MM-DD)
   * @param {number} filters.min_amount - Minimum amount
   * @param {number} filters.max_amount - Maximum amount
   * @param {Array<string>} filters.categories - Array of categories
   * @param {string} filters.category - Single category
   * @param {string} filters.type - Transaction type (income, expense, transfer)
   * @param {number} filters.account_id - Account ID
   * @param {string} filters.search - Search query
   * @param {Array<string>} filters.tags - Array of tags
   * @param {string} filters.sort_by - Sort field (date, amount, category, type, created_at)
   * @param {string} filters.sort_order - Sort order (asc, desc)
   * @returns {Promise<Object>} - Paginated response { data, pagination }
   */
  getAll: async (filters = {}) => {
    try {
      const params = new URLSearchParams();

      // Pagination parameters
      if (filters.page) params.append('page', filters.page);
      if (filters.page_size) params.append('page_size', filters.page_size);

      // Date filters
      if (filters.start_date) params.append('start_date', filters.start_date);
      if (filters.end_date) params.append('end_date', filters.end_date);

      // Amount filters
      if (filters.min_amount !== undefined) params.append('min_amount', filters.min_amount);
      if (filters.max_amount !== undefined) params.append('max_amount', filters.max_amount);

      // Category filters (support both single and multiple)
      if (filters.categories?.length) {
        filters.categories.forEach(cat => params.append('categories', cat));
      } else if (filters.category) {
        params.append('category', filters.category);
      }

      // Type filter
      if (filters.type) params.append('type', filters.type);

      // Account filter
      if (filters.account_id) params.append('account_id', filters.account_id);

      // Search query
      if (filters.search) params.append('search', filters.search);

      // Tags filter
      if (filters.tags?.length) {
        filters.tags.forEach(tag => params.append('tags', tag));
      }

      // Sorting
      if (filters.sort_by) params.append('sort_by', filters.sort_by);
      if (filters.sort_order) params.append('sort_order', filters.sort_order);

      // Gin backend endpoint for transactions with pagination
      const response = await httpClient.backend.get(`/transactions?${params.toString()}`);

      // Handle both paginated and non-paginated responses for backward compatibility
      if (response.data && Array.isArray(response.data)) {
        return {
          data: response.data.map(transaction => Transaction.fromJSON(transaction)),
          total: response.total || response.data.length,
          page: response.page || 1,
          page_size: response.page_size || response.data.length,
          total_pages: response.total_pages || 1
        };
      }

      // Legacy format: direct array response
      if (Array.isArray(response)) {
        return {
          data: response.map(transaction => Transaction.fromJSON(transaction)),
          total: response.length,
          page: 1,
          page_size: response.length,
          total_pages: 1
        };
      }

      throw new Error('Unexpected response format');
    } catch (error) {
      console.error('Error fetching transactions:', error);
      throw error;
    }
  },
  
  /**
   * Get transaction by ID
   * @param {string} id - Transaction ID
   * @returns {Promise<Object>} - Transaction data
   */
  getById: async (id) => {
    try {
      // Gin backend endpoint for single transaction
      const data = await httpClient.backend.get(`/transactions/${id}`);
      return Transaction.fromJSON(data);
    } catch (error) {
      console.error(`Error fetching transaction ${id}:`, error);
      throw error;
    }
  },
  
  /**
   * Create new transaction
   * @param {Object} transactionData - Transaction data
   * @returns {Promise<Object>} - Created transaction
   */
  create: async (transactionData) => {
    try {
      const data = await httpClient.backend.post('/transactions', transactionData);
      return Transaction.fromJSON(data);
    } catch (error) {
      console.error('Error creating transaction:', error);
      throw error;
    }
  },
  
  /**
   * Update transaction
   * @param {string} id - Transaction ID
   * @param {Object} transactionData - Updated transaction data
   * @returns {Promise<Object>} - Updated transaction
   */
  update: async (id, transactionData) => {
    try {
      const data = await httpClient.backend.put(`/transactions/${id}`, transactionData);
      return Transaction.fromJSON(data);
    } catch (error) {
      console.error(`Error updating transaction ${id}:`, error);
      throw error;
    }
  },
  
  /**
   * Delete transaction
   * @param {string} id - Transaction ID
   * @returns {Promise<Object>} - Response data
   */
  delete: async (id) => {
    try {
      return await httpClient.backend.delete(`/transactions/${id}`);
    } catch (error) {
      console.error(`Error deleting transaction ${id}:`, error);
      throw error;
    }
  },
  
  /**
   * Get transaction categories
   * @returns {Promise<Array>} - List of categories
   */
  getCategories: async () => {
    try {
      return await httpClient.backend.get('/transactions/categories');
    } catch (error) {
      console.error('Error fetching transaction categories:', error);
      throw error;
    }
  },
  
  /**
   * Get transaction summary
   * @param {string} period - Time period (week, month, year)
   * @returns {Promise<Object>} - Summary data
   */
  getSummary: async (period = 'month') => {
    try {
      // Gin backend endpoint for transaction summary
      return await httpClient.backend.get(`/transactions/summary?period=${period}`);
    } catch (error) {
      console.error(`Error fetching transaction summary for ${period}:`, error);
      throw error;
    }
  },

  /**
   * Transfer money between accounts
   * @param {Object} transferData - Transfer details
   * @param {number} transferData.amount - Amount to transfer
   * @param {string} transferData.description - Transfer description
   * @param {number} transferData.from_account_id - Source account ID
   * @param {number} transferData.to_account_id - Destination account ID
   * @param {string} transferData.date - Transfer date
   * @param {Array} transferData.tags - Optional tags
   * @returns {Promise<Object>} - Transfer response with both transactions
   */
  transfer: async (transferData) => {
    try {
      return await httpClient.backend.post('/transactions/transfer', transferData);
    } catch (error) {
      console.error('Error transferring money:', error);
      throw error;
    }
  },

  /**
   * Export transactions to CSV
   * @param {Object} filters - Optional filters for export
   * @returns {Promise<Blob>} - CSV file as blob
   */
  exportToCSV: async (filters = {}) => {
    try {
      const queryParams = new URLSearchParams(filters).toString();
      const url = `/transactions/export${queryParams ? `?${queryParams}` : ''}`;
      return await httpClient.backend.get(url, { responseType: 'blob' });
    } catch (error) {
      console.error('Error exporting transactions:', error);
      throw error;
    }
  }
};

export default transactionService;
